nativesoundsystem/src/main/jni/aaudio/AAudioManager.cpp

```diff
@@ -18,74 +18,96 @@
 
-    int32_t framesPerBurst = AAudioStream_getFramesPerBurst(eng->playStream_);
-    int32_t channelCount = AAudioStream_getSamplesPerFrame(eng->playStream_);
+    if (eng->playAudio_) {
+        PrintAudioStreamInfo(engine.playStream_);
 
-    LOGI("channel count %d", channelCount);
-    LOGI("framesPerBurst %d", framesPerBurst);
+        aaudio_stream_state_t aaudioStreamState = AAudioStream_getState(eng->playStream_);
+        LOGV("[AAudioManager] currentState=%d %s", aaudioStreamState,
+             AAudio_convertStreamStateToText(aaudioStreamState));
+//        if (aaudioStreamState == AAUDIO_STREAM_STATE_STARTING) {
+//            memset(static_cast<AUDIO_HARDWARE_SAMPLE_TYPE *>(audioData), 0,
+//                   sizeof(AUDIO_HARDWARE_SAMPLE_TYPE) * samplesPerFrame * numFrames);
+//            return AAUDIO_CALLBACK_RESULT_CONTINUE;
+//        }
 
-    int numberElements = framesPerBurst * channelCount;
-    int sizeBuffer = numberElements * sizeof(AUDIO_HARDWARE_SAMPLE_TYPE);
+        int32_t framesPerBurst = AAudioStream_getFramesPerBurst(eng->playStream_);
+        int32_t channelCount = AAudioStream_getSamplesPerFrame(eng->playStream_);
 
-    AUDIO_HARDWARE_SAMPLE_TYPE *buf = new AUDIO_HARDWARE_SAMPLE_TYPE[sizeBuffer];
-    assert(buf);
+        LOGI("[AAudioManager] channel count %d", channelCount);
+        LOGI("[AAudioManager] framesPerBurst %d", framesPerBurst);
 
-    aaudio_result_t result;
+        int numberElements = numFrames * channelCount;
+        size_t sizeBuffer = numberElements * sizeof(AUDIO_HARDWARE_SAMPLE_TYPE);
+        aaudio_result_t framesWritten;
 
-    while (!eng->requestStop_) {
         if (eng->playAudio_ && eng->playPosition_ < eng->soundSystem_->getTotalNumberFrames()) {
+
+            // End of track
             if (eng->playPosition_ > eng->soundSystem_->getTotalNumberFrames() - numberElements) {
-                sizeBuffer = (eng->soundSystem_->getTotalNumberFrames() - eng->playPosition_) * sizeof(AUDIO_HARDWARE_SAMPLE_TYPE);
+                sizeBuffer = (eng->soundSystem_->getTotalNumberFrames() - eng->playPosition_) *
+                             sizeof(AUDIO_HARDWARE_SAMPLE_TYPE);
             }
-            memmove(buf, eng->soundSystem_->getExtractedData() + eng->playPosition_,
+            memmove(bufferDest, eng->soundSystem_->getExtractedData() + eng->playPosition_,
                     sizeBuffer);
             eng->playPosition_ += numberElements;
         } else {
-            memset(buf, 0, sizeBuffer);
+            memset(bufferDest, 0, sizeBuffer);
         }
-        result = AAudioStream_write(eng->playStream_,
-                                    buf,
-                                    framesPerBurst,
-                                    100000000);
-        assert(result > 0);
-    }
-
-    delete[] buf;
-    eng->requestStop_ = false;
 
-    AAudioStream_requestStop(eng->playStream_);
+        framesWritten = AAudioStream_write(
```

Comment:
AAudioStream_write is become useless when you use AAudio callback. You just have to put data inside buffer from parameters to play sound.